dora/core/views.py

```diff
@@ -37,3 +48,105 @@ def ping(request):
 def trigger_error(request):
     division_by_zero = 1 / 0
     print(division_by_zero)
+
+
+@api_view()
+@permission_classes([permissions.AllowAny])
+def get_inclusion_connect_login_info(request):
+    # https://security.stackexchange.com/questions/147529/openid-connect-nonce-replay-attack
+    # https://stackoverflow.com/questions/46844285/difference-between-oauth-2-0-state-and-openid-nonce-parameter-why-state-cou
+    # https://blogs.aaddevsup.xyz/2019/11/state-parameter-in-mvc-application/
+    # https://stackoverflow.com/questions/35165793/what-attack-does-the-state-parameter-in-openid-connect-server-flow-prevent
+    # https://stackoverflow.com/questions/53246830/how-nonce-and-state-parameters-are-stored-and-transmitted-in-identityserver4
+    # https://stackoverflow.com/questions/46844285/difference-between-oauth-2-0-state-and-openid-nonce-parameter-why-state-cou/46859861#46859861
+    state = get_random_string(32)
+    nonce = get_random_string(32)
+    # Not sure hashing gets us anything…
+    hashed_nonce = make_password(nonce)
+
+    cache.set(
+        f"oidc-state-{state}",
+        {
+            "state": state,
+            "nonce": nonce,
+        },
+    )
+    return Response(
+        {
+            "url": f"{settings.IC_BASE_URL}auth?response_type=code&from=dora&client_id={settings.IC_CLIENT_ID}&scope=openid profile email&nonce={hashed_nonce}",
```

Comment:
Je ne suis pas très fan des constructions d'URL par concaténation. Plusieurs soucis potentiels ici :
 * il n'y a aucune garantie que l'URL de login Inclusion Connect soie (ou reste) une URL se terminant par `auth`. En fait, bien que les différents _endpoints_ OIDC d'un provider donné (ici implémenté par Keycloak) puissent se ressembler, la spec ne dit rien là-dessus et donc l'approche à base de `IC_BASE_URL` est fragile. Les options sont essentiellement : soit avoir une variable de configuration par endpoint, donc par exemple `IC_AUTH_URL`, `IC_TOKEN_URL` et `IC_LOGOUT_URL` (aussi `IC_USERINFO_URL` si c'était utile), soit utiliser le mécanisme de "discovery" de OIDC (https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderConfigurationRequest) auquel cas on n'aurait besoin que d'une seule variable par exemple `IC_CONFIG_URL` mais ça nécessiterait d'écrire du code pour récupérer les différentes URL en téléchargeant le JSON en question (ou d'utiliser une lib client OIDC, qui sait le faire en principe).
 * la construction par concaténation mène à plein de problèmes d'échappement ; par exemple ici, est-ce que le _client id_ ne contient pas de caractères comme `&` qui nécessiteraient un échappement ? Qu'en est-il du `hashed_nonce` ? Et les espaces dans `scope` ne sont pas techniquement légaux dans une URL.

Ma suggestion est de construire plutôt cette URL en utilisant `urllib.parse`, par exemple `urllib.parse.urlencode({'response_type':'code', 'from':'dora', 'client_id':settings.IC_CLIENT_ID, …})`

Reply:
- bien vu pour une variable de config par endpoint.
- d'accord avec toi pour la construction par concaténation; pour tous les paramètres que tu cite, on pourrait argumenter que je les controle et qu'il n'y a donc pas de problème, mais je n'encode même pas le redirect_uri coté front pour l'instant; j'avais sur ma todo list de le passer dans un encodeURIComponent, mais tout passer par urllib.parse sera effectivement plus propre.

Merci encore!!!

Reply:
J'ajoute une subtilité sur la construction d'URL, qui oblige à faire vraiment proprement : la spec OAuth 2.0 (https://www.rfc-editor.org/rfc/rfc6749) dit au sujet de chacun des endpoints (Authorize, Token…) :
> The endpoint URI MAY include an "application/x-www-form-urlencoded"
   formatted (per [Appendix B](https://www.rfc-editor.org/rfc/rfc6749#appendix-B)) query component ([[RFC3986] Section 3.4](https://www.rfc-editor.org/rfc/rfc3986#section-3.4)),
   which MUST be retained when adding additional query parameters.

Autrement dit si ton provider OIDC dit que son "Authorization endpoint" est `https://provider.example.net/auth?x=1&y=2` alors il faut que tu construises l'URL `https://provider.example.net/auth?x=1&y=2&response_type=code&client_id=…`. Mais bon, avec une bonne lib de traitement d'URL ça se fait assez facilement (parser l'URL fournie, ajouter des entrées au dictionnaire des queryparams, re-générer l'URL).